File: SCAHomework2/Plants/container.cpp
//
// Created by Иван Доронин on 23.10.2021.
//
//------------------------------------------------------------------------------
// container.cpp - содержит функции обработки контейнера
//------------------------------------------------------------------------------
#include "container.h"
//------------------------------------------------------------------------------
// Очистка контейнера от элементов (отчистка памяти).
void Container::Clear() {
    for (int i = 0; i < length; ++i) {
        delete data[i];
    }
    delete data;
    length = 0;
}
//------------------------------------------------------------------------------
// Ввод содержимого контейнера из указанного потока.
void Container::In(FILE *file, const int *size) {
    int result, plant_type;
    while(length < *size) {
        result = 1;
        if (fscanf(file, "%d", &plant_type) == EOF) {
            return;
        }
        switch(plant_type) {
            case 1:
                data[length] = new Tree();
                break;
            case 2:
                data[length] = new Flower();
                break;
            case 3:
                data[length] = new Shrub();
                break;
            default:
                result = 0;
        }
        result = (result) ? data[length]->In(file) : 0;

        if (result == 1) {
            length++;
        }
        else if (result == 2) {
            return;
        } else {
            delete data[length];
        }
    }
}
//------------------------------------------------------------------------------
// Случайный ввод содержимого контейнера.
void Container::InRnd(int size) {
    printf("Start random filling: \n");
    while(length < size) {
        switch (rand() % 3) {
            case 0:
                data[length] = new Tree();
                break;
            case 1:
                data[length] = new Flower();
                break;
            case 2:
                data[length] = new Shrub();
                break;
        }
        data[length++]->InRnd();
    }
    printf("End random filling.\n");
}
//------------------------------------------------------------------------------
// Инициализация контейнера.
void Container::Init(int *size) {
    printf("Start init: \n");
    data = new Plant*[*size];
    length = 0;
    printf("End init\n");
}
//------------------------------------------------------------------------------
// Вывод содержимого контейнера в указанный файл.
void Container::Out(FILE *fp) {
    fprintf(fp, "Container contains %d elements.\n", length);
    for(int i = 0; i < length; i++) {
        fprintf(fp, "%d: ", i);
        data[i]->Out(fp);
    }
}
//------------------------------------------------------------------------------
// Меняем местами элементы.
template<typename T>
void swap(T &a, T &b) {
    T temp(std::move(a));
    a = std::move(b);
    b = std::move(temp);
}
// Определяем индексы.
void Container::Heapify(int x, int y) {
    int largest = y;
    int l = 2 * y + 1;
    int r = 2 * y + 2;

    if (l < x && data[l]->ComparisonParam() > data[largest]->ComparisonParam())largest = l;
    if (r < x && data[r]->ComparisonParam() > data[largest]->ComparisonParam())largest = r;

    if (largest != y) {
        swap(data[y], data[largest]);
        Heapify(x, largest);
    }
}
// Сортировка.
void Container::HeapSort() {
    printf("Start heap sorting: \n");
    int n = length;
    for (int i = n / 2 - 1; i >= 0; i--)
        Heapify(n, i);

    for (int i = n - 1; i >= 0; i--) {
        swap(data[0], data[i]);
        Heapify(i, 0);
    }
    printf("End heap sorting.\n");
}

File: pythonCSAHW3/container.py
# ------------------------------------------------------------------------------
# container.py - Container handling functions.
# ------------------------------------------------------------------------------

import random
from extender import *


# ------------------------------------------------------------------------------
# Input data from file.
def in_from_file(cont, plants_data):
    for plant in plants_data:
        type_of_plant = plant[0]
        if type_of_plant == '1':
            tree_arr = []
            tree.in_from_file(tree_arr, plant)
            cont.append(tree_arr)
        if type_of_plant == '2':
            flower_arr = []
            flower.in_from_file(flower_arr, plant)
            cont.append(flower_arr)
        if type_of_plant == '3':
            shrub_arr = []
            shrub.in_from_file(shrub_arr, plant)
            cont.append(shrub_arr)


# ------------------------------------------------------------------------------
# Random filling container.
def in_rnd(cont, size):
    for _ in range(size):
        type_of_plant = random.randint(1, 3)
        if type_of_plant == 1:
            tree_arr = []
            tree.in_rnd(tree_arr)
            cont.append(tree_arr)
        if type_of_plant == 2:
            flower_arr = []
            flower.in_rnd(flower_arr)
            cont.append(flower_arr)
        if type_of_plant == 3:
            shrub_arr = []
            shrub.in_rnd(shrub_arr)
            cont.append(shrub_arr)


# ------------------------------------------------------------------------------
# Output in file.
def out(cont, out_stream):
    out_stream.write(f"Container contains {len(cont)} elements:\n")
    count = 0
    for plant in cont:
        count += 1
        out_stream.write(f"{count}: It is a {plant[0]}, ")

        if plant[0] == "Tree":
            out_stream.write(f"name = {plant[1]}, age = {plant[2]}, ")
        elif plant[0] == "Shrub":
            out_stream.write(f"name = {plant[1]}, flowering month = {plant[2]}, ")
        elif plant[0] == "Flower":
            out_stream.write(f"name = {plant[1]}, type of flower = {plant[2]}")

        out_stream.write(f" comparison parameter = {plant[len(plant) - 1]}\n")


# ------------------------------------------------------------------------------
# Helpful method for heap_sort.
def heapify(plant_arr, n, i):
    largest = i  # Initialize largest as root
    l = 2 * i + 1  # left = 2*i + 1
    r = 2 * i + 2  # right = 2*i + 2

    # See if left child of root exists and is
    # greater than root
    if l < n and plant_arr[i][3] < plant_arr[l][3]:
        largest = l

    # See if right child of root exists and is
    # greater than root
    if r < n and plant_arr[largest][3] < plant_arr[r][3]:
        largest = r

    # Change root, if needed
    if largest != i:
        plant_arr[i], plant_arr[largest] = plant_arr[largest], plant_arr[i]  # swap

        # Heapify the root.
        heapify(plant_arr, n, largest)


# ------------------------------------------------------------------------------
# Heap sorting.
def heap_sort(cont):
    n = len(cont)

    for i in range(n // 2 - 1, -1, -1):
        heapify(cont, n, i)

    # One by one extract elements
    for i in range(n - 1, 0, -1):
        cont[i], cont[0] = cont[0], cont[i]
        heapify(cont, i, 0)